### PR TITLE
#1148: Support updates to Instances

### LINKED ISSUE #1148: Proposal: Plan and Parameter updates
### Proposal: plan and parameter updates

The user should be able to update the plan and parameters in the `ServiceInstance` resource to cause the instance to be updated in the broker.

#### Secret-parameter-only changes

The `ServiceInstance` object allows users to reference secrets from which parameters are sourced.

There are two ways a user might choose to do this:

1.  Update the instance resource with a new secret name
2.  Update the secret out of band

In the case where a new secret is used, the controller can detect the spec change using the generation field.

In our discussions on this problem so far, we have talked most about an out-of-band change to the secret initially used to supply parameters.

The controller is not allowed to watch secrets.  There must be a way for users to indicate that updates for secret parameters have changed out of band without a spec change in the object.

We can limit unexpected behaviors in this resource by adopting a model out-of-band changes to secrets containing parameters after they are initially used have no effect.  So, I wonder from the group, does anyone actually have a specific use-case that requires out-of-band update without changing the secret name?

We should consider adding support for out of band updates later.

If we have to support them, we could use a field similar to the `RelistRequests` to drive a new generation of the resource from.  If the user needs out of band updates to secrests but doesn't want to change the plan, they can bump the `UpdateRequests` field of the InstanceSpec.

#### API Changes

```go
// InstanceSpec represents the desired state of an Instance.
type InstanceSpec struct {
	// ServiceClassName is the name of the ServiceClass this Instance
	// should be provisioned from.
	//
	// Immutable.
	ServiceClassName string

	// PlanName is the name of the ServicePlan this Instance should be
	// provisioned from.  Mutable.
	PlanName string

	// Parameters is a set of the parameters to be
	// passed to the underlying broker.
	// The inline YAML/JSON payload to be translated into equivalent
	// JSON object.
	// If a top-level parameter name exists in multiples sources among
	// `Parameters` and `ParametersFrom` fields, it is
	// considered to be a user error in the specification.
	// Mutable.
	// +optional
	Parameters *runtime.RawExtension

	// List of sources to populate parameters.
	// If a top-level parameter name exists in multiples sources among
	// `Parameters` and `ParametersFrom` fields, it is
	// considered to be a user error in the specification.
	// Mutable.
	// +optional
	ParametersFrom []ParametersFromSource

	// ExternalID is the identity of this object for use with the OSB API.
	//
	// Immutable.
	ExternalID string
}
```

### PR DESCRIPTION
**Pull request by staebler**
Closes #1148, #816, #748, #1062, and #1065.

* Allow parameters and plan to be changed for a ServiceInstance.
* Send an UpdateInstanceRequest to the Broker when the spec changes.
* Add UpdateRequests field to ServiceInstance spec that can be bumped to provoke a new UpdateInstanceRequest to the Broker.

### CODE REVIEW

**Comment by MHBauer** [comment]
`go vet` errors need fixing.

**Comment by pmorie** [comment]
@staebler rebase needed

File: pkg/apis/servicecatalog/types.go [hunk 1]
```diff
@@ -445,6 +439,12 @@ type ServiceInstanceSpec struct {
 	// end-user. User-provided values for this field are not saved.
 	// +optional
 	UserInfo *UserInfo
+
+	// UpdateRequests is a strictly increasing, non-negative integer counter that
+	// can be manually incremented by a user to manually trigger an update. This
+	// allows for parameters to be updated with any out-of-band changes that have
+	// been made to the secrets from which the parameters are sourced.
+	UpdateRequests int64
```
Comment (nilebox) [inline pkg/apis/servicecatalog/types.go]: if it's non-negative as the comment says, shall we use the `uint64` type?

File: pkg/apis/servicecatalog/validation/instance.go [hunk 2]
```diff
@@ -234,8 +236,17 @@ func internalValidateServiceInstanceUpdateAllowed(new *sc.ServiceInstance, old *
 // ValidateServiceInstanceUpdate validates a change to the Instance's spec.
 func ValidateServiceInstanceUpdate(new *sc.ServiceInstance, old *sc.ServiceInstance) field.ErrorList {
 	allErrs := field.ErrorList{}
+
 	allErrs = append(allErrs, internalValidateServiceInstanceUpdateAllowed(new, old)...)
 	allErrs = append(allErrs, internalValidateServiceInstance(new, false)...)
+
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(new.Spec.ExternalServiceClassName, old.Spec.ExternalServiceClassName, field.NewPath("spec").Child("externalServiceClassName"))...)
```
Comment (nilebox) [inline pkg/apis/servicecatalog/validation/instance.go]: We also need to check that plan hasn't changed if plan updates are not supported (`ServiceClass.Spec.PlanUpdatable`==false).
If it's costly to validate it there, we can check it before sending an update request to the broker.

File: pkg/controller/controller_instance.go [hunk 3]
```diff
@@ -644,30 +628,97 @@ func (c *controller) reconcileServiceInstance(instance *v1alpha1.ServiceInstance
...
+		updateRequest = &osb.UpdateInstanceRequest{
+			AcceptsIncomplete:   true,
+			InstanceID:          instance.Spec.ExternalID,
+			ServiceID:           serviceClass.Spec.ExternalID,
+			OriginatingIdentity: originatingIdentity,
+		}
+		// Only send the plan ID if the plan name has changed from what the Broker has
+		if toUpdate.Status.ExternalProperties == nil ||
+			toUpdate.Status.InProgressProperties.ExternalServicePlanName != toUpdate.Status.ExternalProperties.ExternalServicePlanName {
+			planID := servicePlan.Spec.ExternalID
```
Comment (nilebox) [inline pkg/controller/controller_instance.go]: We need to check the flag `plan_updatable` in `ServiceClass.Spec.PlanUpdatable`

File: pkg/apis/servicecatalog/types.go (hunk 1 above)
Comment (staebler) [reply pkg/apis/servicecatalog/types.go]: From https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#primitive-types

> Do not use unsigned integers, due to inconsistent support across languages and libraries. Just validate that the integer is non-negative if that's the case.

File: pkg/apis/servicecatalog/validation/instance.go (hunk 2 above)
Comment (staebler) [reply pkg/apis/servicecatalog/validation/instance.go]: I am not sure about this, but it is my understanding that the validation should not be looking at other resources. The validation should only consider the resource being validated. There is an admission controller [1] that blocks updates to `ExternalServicePlanName` when the `PlanUpdatable` field of the service class is false.

[1] https://github.com/kubernetes-incubator/service-catalog/blob/master/plugin/pkg/admission/serviceplan/changevalidator/admission.go

File: pkg/controller/controller_instance.go (hunk 3 above)
Comment (staebler) [reply pkg/controller/controller_instance.go]: I don't think that this is the right place to check this. The change to the plan name has already been accepted by the API Server, given the state of the resources at that time. I don't think it is appropriate to ignore the accepted change to the plan name given the state of the resources at this time. We would need to do more than just not send the plan change to the Broker if that were the case. We would also need to somehow revert the plan change in the Spec of the ServiceInstance.

File: pkg/apis/servicecatalog/types.go (hunk 1 above)
Comment (nilebox) [reply pkg/apis/servicecatalog/types.go]: thanks

File: pkg/apis/servicecatalog/validation/instance.go (hunk 2 above)
Comment (nilebox) [reply pkg/apis/servicecatalog/validation/instance.go]: @staebler didn't know we have an admission controller for that. then there is no need for checking it again, of course.

**Comment by vaikas-google** [comment]
This has nothing to do with your PR, nicely injected in there and from diffs easy to reason about. But, we really have to do some refactoring here, having a method that's several hundred lines intertwined with all kinds of different ifs sprinkled in all over makes it really really hard to digest what's going on.

**Comment by pmorie** [comment]
I am fully supportive of refactoring after all renames and functional
features are done :)

**Comment by vaikas-google** [comment]
yes, hence LGTM

**Comment by staebler** [comment]
I also whole-heartedly support refactoring.

File: pkg/controller/controller_instance.go [hunk 4]
```diff
@@ -687,8 +738,8 @@ func (c *controller) reconcileServiceInstance(instance *v1alpha1.ServiceInstance
 				toUpdate,
 				v1alpha1.ServiceInstanceConditionReady,
 				v1alpha1.ConditionFalse,
-				errorProvisionCallFailedReason,
-				"ClusterServiceBroker returned a failure for provision call; operation will not be retried: "+s)
+				reason,
+				fmt.Sprintf("ClusterServiceBroker returned a failure for %v call; operation will not be retried: %v", provisionOrUpdateText, s))
 
 			if shouldStartOrphanMitigation(httpErr.StatusCode) {
```
Comment (pmorie) [inline pkg/controller/controller_instance.go]: You should only do orphan mitigation here if you're provisioning, not if you're updating.

File: pkg/registry/servicecatalog/instance/strategy.go [hunk 5]
```diff
@@ -153,23 +153,19 @@ func (instanceRESTStrategy) PrepareForUpdate(ctx genericapirequest.Context, new,
...
 	// spec changes and other changes to the object.
-	//
-	// Note that since we do not currently handle any changes to the spec,
-	// the generation will never be incremented
 	if !apiequality.Semantic.DeepEqual(oldServiceInstance.Spec, newServiceInstance.Spec) {
 		if utilfeature.DefaultFeatureGate.Enabled(scfeatures.OriginatingIdentity) {
 			setServiceInstanceUserInfo(newServiceInstance, ctx)
 		}
 		newServiceInstance.Generation = oldServiceInstance.Generation + 1
+		setServiceInstanceReadyFalseCondition(newServiceInstance)
```
Comment (pmorie) [inline pkg/registry/servicecatalog/instance/strategy.go]: In a follow-up, we need to take this out.  I'm sorry that I missed it earlier, but update should _not_ affect the status of a resource. It's okay that the ready condition remains set after this - the controller will eventually unset it. Remember, the system is not transactional and only has to eventually converge on the right state.

**Comment by pmorie** [comment]
LGTM but we need to correct the status setting in the API server in a follow-up.

**Pull request merged**